File: power-adapters/src/test/java/com/nextfaze/poweradapters/FakeAdapter.java
package com.nextfaze.poweradapters;

import android.support.annotation.NonNull;
import android.view.View;
import android.view.ViewGroup;
import com.nextfaze.poweradapters.internal.NotificationType;

import static com.nextfaze.poweradapters.internal.Preconditions.checkNotNull;

@SuppressWarnings("WeakerAccess")
public class FakeAdapter extends PowerAdapter {

    private int mItemCount;

    @NonNull
    private NotificationType mNotificationType = NotificationType.FINE;

    public FakeAdapter(int itemCount) {
        mItemCount = itemCount;
    }

    @NonNull
    @Override
    public Object getItemViewType(int position) {
        assertWithinRange(position);
        return super.getItemViewType(position);
    }

    @Override
    public long getItemId(int position) {
        assertWithinRange(position);
        return NO_ID;
    }

    @Override
    public boolean isEnabled(int position) {
        assertWithinRange(position);
        return true;
    }

    @NonNull
    @Override
    public View newView(@NonNull ViewGroup parent, @NonNull Object viewType) {
        return new View(parent.getContext());
    }

    @Override
    public void bindView(@NonNull Container container, @NonNull View view, @NonNull Holder holder) {
    }

    @Override
    public boolean hasStableIds() {
        return false;
    }

    @Override
    public int getItemCount() {
        return mItemCount;
    }

    public void insert(int positionStart, int itemCount) {
        mItemCount += itemCount;
        mNotificationType.notifyItemRangeInserted(mDataObservable, positionStart, itemCount);
    }

    public void append(int itemCount) {
        insert(mItemCount, itemCount);
    }

    public void remove(int positionStart, int itemCount) {
        if (positionStart + itemCount > mItemCount) {
            throw new IndexOutOfBoundsException();
        }
        mItemCount -= itemCount;
        mNotificationType.notifyItemRangeRemoved(mDataObservable, positionStart, itemCount);
    }

    public void change(int positionStart, int itemCount) {
        mNotificationType.notifyItemRangeChanged(mDataObservable, positionStart, itemCount);
    }

    public void move(int fromPosition, int toPosition, int itemCount) {
        mNotificationType.notifyItemRangeMoved(mDataObservable, fromPosition, toPosition, itemCount);
    }

    public void clear() {
        remove(0, mItemCount);
    }

    @SuppressWarnings("unused")
    @NonNull
    public NotificationType getNotificationType() {
        return mNotificationType;
    }

    public void setNotificationType(@NonNull NotificationType notificationType) {
        mNotificationType = checkNotNull(notificationType, "notificationType");
    }

    private void assertWithinRange(int position) {
        if (position >= mItemCount) {
            throw new IndexOutOfBoundsException();
        }
    }
}

File: power-adapters-test/src/main/java/com/nextfaze/poweradapters/test/FakeAdapter.kt
package com.nextfaze.poweradapters.test

import android.view.View
import android.view.ViewGroup
import com.nextfaze.poweradapters.Container
import com.nextfaze.poweradapters.Holder
import com.nextfaze.poweradapters.PowerAdapter
import com.nextfaze.poweradapters.internal.NotificationType

class FakeAdapter @JvmOverloads constructor(private var itemCount: Int = 0) : PowerAdapter() {

    var notificationType = NotificationType.FINE

    override fun getItemViewType(position: Int): Any {
        assertWithinRange(position)
        return super.getItemViewType(position)
    }

    override fun getItemId(position: Int): Long {
        assertWithinRange(position)
        return PowerAdapter.NO_ID.toLong()
    }

    override fun isEnabled(position: Int): Boolean {
        assertWithinRange(position)
        return true
    }

    override fun newView(parent: ViewGroup, viewType: Any) = View(parent.context)

    override fun bindView(container: Container, view: View, holder: Holder) {
    }

    override fun hasStableIds() = false

    override fun getItemCount() = itemCount

    fun insert(positionStart: Int, itemCount: Int) {
        this.itemCount += itemCount
        notificationType.notifyItemRangeInserted(dataObservable, positionStart, itemCount)
    }

    fun append(itemCount: Int) {
        insert(this.itemCount, itemCount)
    }

    fun remove(positionStart: Int, itemCount: Int) {
        if (positionStart + itemCount > this.itemCount) {
            throw IndexOutOfBoundsException()
        }
        this.itemCount -= itemCount
        notificationType.notifyItemRangeRemoved(dataObservable, positionStart, itemCount)
    }

    fun change(positionStart: Int, itemCount: Int) {
        notificationType.notifyItemRangeChanged(dataObservable, positionStart, itemCount)
    }

    fun move(fromPosition: Int, toPosition: Int, itemCount: Int) {
        notificationType.notifyItemRangeMoved(dataObservable, fromPosition, toPosition, itemCount)
    }

    fun clear() = remove(0, itemCount)

    private fun assertWithinRange(position: Int) {
        if (position >= itemCount) throw IndexOutOfBoundsException()
    }
}
